lib/manager.js

```diff
@@ -0,0 +1,93 @@
+"use strict";
+
+let Manifest = require("./manifest");
+let { createFile, generateFingerprint } = require("faucet-pipeline-util");
+let path = require("path");
+
+module.exports = class AssetManager {
+	constructor(referenceDir, { manifestConfig, fingerprint, exitOnError } = {}) {
+		this.referenceDir = referenceDir;
+		this.fingerprint = fingerprint;
+		this.exitOnError = exitOnError;
+		let manifest = this.resolvePath(manifestConfig.file, { enforceRelative: true });
+		this.manifest = new Manifest(manifest, manifestConfig.baseURI);
+
+		// bind methods for convenience
+		this.writeFile = this.writeFile.bind(this);
+		this.resolvePath = this.resolvePath.bind(this);
+	}
+
+	writeFile(filepath, data, { error } = {}) {
+		let originalPath = filepath;
+		if(this.fingerprint) {
+			filepath = generateFingerprint(filepath, data);
+		}
+
+		return createFile(filepath, data).
+			then(_ => this._updateManifest(originalPath, filepath)).
+			then(_ => {
+				this._report(originalPath, error);
+				if(error && this.exitOnError) {
+					throw error;
+				}
+			}).
+			catch(err => {
+				console.error("aborting");
+				process.exit(1);
+			});
+	}
+
+	resolvePath(filepath, { enforceRelative } = {}) {
+		if(filepath.substr(0, 2) === "./") {
+			return path.resolve(this.referenceDir, filepath);
+		} else if(enforceRelative) {
+			throw new Error(`path must be relative: \`${repr(filepath)}\``);
+		} else { // attempt via Node resolution algorithm
+			try {
+				return resolveModulePath(filepath, this.referenceDir);
+			} catch(err) {
+				throw new Error(`could not resolve \`${repr(filepath)}\``);
+			}
+		}
+	}
+
+	_updateManifest(originalPath, actualPath) {
+		let { manifest, referenceDir } = this;
+		originalPath = path.relative(referenceDir, originalPath);
+		actualPath = path.relative(referenceDir, actualPath);
+		return manifest.set(originalPath, actualPath);
```

Comment:
re https://github.com/faucet-pipeline/faucet-pipeline/pull/17#discussion_r159960476 (GitHub won't let me comment on outdated diffs as part of a review batch):

Yeah, looks like we had pretty much the same idea there: https://github.com/faucet-pipeline/faucet-pipeline/pull/17#discussion_r159902086

> TBQH I would move this entire _updateManifest method to the manifest, it has exactly the signature that I would like the add method to have.

Good point, addressed (mostly) in 72be16839fb0a9dc35a3c4146214ffdf6bdf6fda.

Reply:
Why do you think that the responsibility of making the paths relative to the referenceDir are the responsibility of AssetManager and not Manifest?

Reply:
Because handling file-system issues is `AssetManagers` shtick; `Manifest` should only concern itself with maintaining the mapping (and, admittedly, the URI - but while similar on the surface, semantically that's different, which is why it's defined in `manifestConfig`).

Reply:
Hm... I'm not entirely convinced, but I think you will change your mind when we implement the changes I proposed to the manifest generation.

Reply:
I'd be [happy to be convinced otherwise](http://bobsutton.typepad.com/my_weblog/2006/07/strong_opinions.html).